inference-engine/src/transformations/src/transformations/serialize.cpp

```diff
@@ -57,10 +57,60 @@ std::string translate_type_name(const std::string& name) {
     return name;
 }
 
+size_t hash_combine(const void* v, int64_t size) {
+    constexpr auto cel_size = sizeof(size_t);
+    size_t seed = static_cast<size_t>(size);
+    const auto data = static_cast<const size_t*>(v);
+    const auto d_end = std::next(data, size / cel_size);
+    for (auto d = data; d != d_end; ++d) {
+        seed ^= *d + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    }
+    size_t last_bytes{0};
+    std::memcpy(&last_bytes, d_end, size % cel_size);
+    seed ^= last_bytes + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    return seed;
+}
+
+class ConstantWriter {
+public:
+    using FilePosition = int64_t;
+    using HashValue = size_t;
+    using ConstWritePositions = std::unordered_map<HashValue, FilePosition>;
+
+    ConstantWriter(std::ostream& bin_data, bool enable_compression = true)
+        : m_binary_output(bin_data)
+        , m_enable_compression(enable_compression) {
+    }
+
+    FilePosition write(const char* ptr, size_t size) {
+        const auto offset = m_binary_output.tellp();
+        if (!m_enable_compression) {
+            m_binary_output.write(ptr, size);
+            return offset;
+        }
+
+        const HashValue hash = hash_combine(ptr, size);
+        const auto found = m_hash_to_file_positions.find(hash);
+        if (found != end(m_hash_to_file_positions)) {
+            return found->second;
+        }
```

Comment:
@blesniewski please explain why you decide that you don't need to expect collisions and resolve them? Is probability of collision is so low that it can be neglected ? If so how you calculated this? By collision i mean the case that two different constants can give the same hash.

Reply:
There is this [article ](https://preshing.com/20110504/hash-collision-probabilities/)for example, you could just take look at the table at the article's end. My thinking here is that we're using 64 bit hashes, for not-that-many constant nodes, so it should be negligible. We're also taking the constant data's size into consideration when creating the hashes (the used hashing algo may not be perfect, but I believe that it suits it's purpose here).

Reply:
Ok. I agree with this approach, but let's be professional here and document this decision in the comment. Keep in mind that MO uses 512bit hash so our approach can be questioned in the future (as I did in this review). 

So to get my approval, please do the following:
- take the largest production model used in validation you can find (probably we can take the .bin file size into account) and count the number of constants in xml, it will give us some concrete hashes number, 'not-that-many' is not so precise :) 
- read out the probability for this amount of hashes and 64bit hash size from the table
- document this calculations and probability in the comment

Reply:
As of this moment, the biggest (in terms of the number of Constant nodes) model I could find, has 2172 constants. I'm going to make an appropriate comment.